services/src/pro/datasets/postgres.rs

```diff
@@ -486,6 +486,39 @@ where
         Ok(id)
     }
 
+    async fn delete_dataset(&self, session: &UserSession, dataset_id: DatasetId) -> Result<()> {
+        let mut conn = self.conn_pool.get().await?;
+
+        let tx = conn.build_transaction().start().await?;
+
+        let stmt = tx
+            .prepare(
+                "
+        SELECT 
+            TRUE
+        FROM 
+            user_permitted_datasets p JOIN datasets d 
+                ON (p.dataset_id = d.id)
+        WHERE 
+            d.id = $1 AND p.user_id = $2 AND p.permission = 'Owner';",
+            )
+            .await?;
+
+        let rows = tx.query(&stmt, &[&dataset_id, &session.user.id]).await?;
+
+        if rows.is_empty() {
+            return Err(Error::OperationRequiresOwnerPermission);
+        }
+
+        let stmt = tx.prepare("DELETE FROM datasets WHERE id = $1;").await?;
```

Comment:
We discussed that we wanted to encapsulate the permission checks from the data management. We said we want to abstract it and that's why I did it here. Also: in order to give a good error message we must do it this way in order to differentiate if the error comes from the permission or from a missing dataset.

Reply:
So it should be a function of its own and at some point it will be part of a separate service?

Reply:
it could be a function of its own, but as we use it only once, I think it can stay there. In the future we will most likely replace all of the rights stuff once we decided on what technology to go with. Everything we do here is only temporary.